Ensure IResult types throw ArgumentNullExceptions on public methods

DamianEdwards opened the pull request:
# Ensure IResult types throw ArgumentNullExceptions on public methods
Updated the `ExecuteAsync(HttpContext context)` and where applicable `PopulateMetadata(EndpointMetadataContext context)` methods of the in-box `IResult` types to throw `ArgumentNullException` if the required method args are `null`.

Added tests to verify.

Fixes #41383

DamianEdwards committed:
Ensure IResult types throw ArgumentNullExceptions on public methods

Fixes #41383

halter73 commented on src/Http/Http.Results/src/Accepted.cs:
```diff
@@ -31,10 +31,7 @@ internal Accepted(string? location)
-        {
-            throw new ArgumentNullException(nameof(locationUri));
-        }
+        ArgumentNullException.ThrowIfNull(locationUri);
```
This seems to be one place where we didn't actually need this since this is internal and we already null check in `TypedResults.Accepted(Uri)`. We don't have any tests for `ArgumentNullExceptions` thrown from the `TypedResults` methods, do we?

DamianEdwards replied:
[We do actually](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Results/test/TypedResultsTests.cs#L97).

I actually think most of these internal constructors are unused and left-over from when the types were being prepared to be made public and have public constructors. I can go through them and remove all the unused constructors

halter73 commented on src/Http/Http.Results/src/ResultsOfT.cs:
```diff
@@ -14,7 +14,7 @@ namespace Microsoft.AspNetCore.Http.HttpResults;
 /// <remarks>
 /// An instance of this type cannot be created explicitly. Use the implicit cast operators to create an instance
 /// from an instance of one of the declared type arguments, e.g.
-/// <code>Results&lt;OkObjectHttpResult, ProblemHttpResult&gt; result = Results.Ok();</code>
```
Can we rename this file to `ResultsOfT.Generated.cs` similar to how we did for `ResultsOfTTests.Generated.cs`?

DamianEdwards committed:
PR feedback & more tests

- Removed unused internal constructors result types
- Ensure argument exceptions  are thrown by public ctors & static factory methods
- Do not throw argument exceptions from internal ctors
